modin/experimental/xgboost/xgboost.py

```diff
@@ -0,0 +1,499 @@
+            node, nthread=nthread
+        )
+        for node in ray.cluster_resources()
+        if "node" in node
+    ]
+
+    assert num_nodes == len(
+        actors
+    ), f"Number of nodes {num_nodes} is not equal to number of actors {len(actors)}."
+
+    return actors
+
+
+def _split_data_across_actors(
+    actors, set_func, X_parts, y_parts=None, evenly_data_distribution=True
+):
+    X_parts_by_actors = _set_row_partitions_by_actors(
+        actors, X_parts, evenly_data_distribution=evenly_data_distribution
+    )
+    if y_parts is not None:
+        y_parts_by_actors = _set_row_partitions_by_actors(
+            actors,
+            y_parts,
+            X_parts_by_actors,
+            evenly_data_distribution=evenly_data_distribution,
+        )
+
+    # Need to add assert on order of actors in case evenly_data_distribution=False
+    for i, actor in enumerate(actors):
+
+        X_parts = (
+            X_parts_by_actors[i]
+            if evenly_data_distribution
+            else X_parts_by_actors[i][2]
+        )
+        if y_parts is None:
+            set_func(actor, *X_parts)
+        else:
+            y_parts = (
+                y_parts_by_actors[i]
+                if evenly_data_distribution
+                else y_parts_by_actors[i][2]
+            )
+            set_func(actor, *(X_parts + y_parts))
+
+
+def _set_row_partitions_by_actors(
+    actors, row_partitions, data_for_aligning=None, evenly_data_distribution=True
+):
+
+    row_partitions_by_actors = []
+    if evenly_data_distribution:
+        num_actors = len(actors)
+        row_parts_last_idx = (
+            len(row_partitions) // num_actors
+            if len(row_partitions) % num_actors == 0
+            else len(row_partitions) // num_actors + 1
+        )
+        start_idx = 0
+        for actor in actors:
+            idx_slice = (
+                slice(start_idx, start_idx + row_parts_last_idx)
+                if start_idx + row_parts_last_idx < len(row_partitions)
+                else slice(start_idx, len(row_partitions))
+            )
+            row_partitions_by_actors.append(row_partitions[idx_slice])
+            start_idx += row_parts_last_idx
+    else:
+        actors_ips = [ray.get(actor.get_actor_ip.remote()) for actor in actors]
+        partitions_ips = [ray.get(row[0]) for row in row_partitions]
+
+        # If one of actors isn't contain partitions we make evenly
+        # distribution of data by nodes.
+        if (
+            len(np.unique(actors_ips)) != len(np.unique(partitions_ips))
+            and data_for_aligning is None
+        ):
+            return _get_evenly_distr_of_partitions_by_ips(
+                actors, actors_ips, row_partitions
+            )
+
+        if data_for_aligning is None:
+            for actor_ip in actors_ips:
+                actor_row_partitions = []
+                order_of_indexes = []
+                for i, row in enumerate(row_partitions):
```

Comment:
this looks like `O(N^2)` algo where using a dict (or an OrderedDict if order matters) should be `O(N*log N)`... dunno if it matters as I don't know how big are actors' and partitions' amounts

Reply:
Number of partitions is more than number of actors. As a rule, partitions' amount is measured from decades to thousands and actors' amount is measured from some to decades/hundreds (in depend of a cluster used).

Reply:
Thank you! Now `dict` is used for set of `actors`, part of loops was changed for decreasing complexity.